### ISSUE TITLE
support editing/adding colour elements in header editor

### ISSUE THREAD

**Issue by dericed** [mbunkus/mkvtoolnix#2038]

**Comment by kieranjol** [comment]
:+1:

**Comment by kieranjol** [comment]
Might as well link to the conversation on Cellar as well: https://mailarchive.ietf.org/arch/search/?email_list=cellar

**Comment by mbunkus** [comment]
I've uploaded new [pre-builds for Windows](https://mkvtoolnix.download/windows/pre/) that include the functionality. I'd appreciate some testing.

**Comment by kieranjol** [comment]
Oh wow, I'll get to this as soon as I can, this is incredible! Thanks
Moritz!

**Comment by kieranjol** [comment]
Hi Moritz,

I was able to add the extra metadata just fine with 
`mkvpropedit.exe --edit track:v1 --set colour-primaries=5 --set colour-range=1 --set colour-matrix=5 --set colour-transfer-characteristics=3 oe5909_vrep_clip.mkv`
using https://drive.google.com/open?id=0B80EiSambHhLSnMwQnc1NFZ0Mnc
as my source file. Thanks! This will make life a lot easier for us in the Irish Film Institute as we can add this metadata to several hundred files without creating new versions. The resulting file now passes the Mediaconch 'is Matroska well described policy' .I have some feedback though;
* According to mediaconch, the file that was generated is not a valid Matroska file. It seems the element in question is:
```
<test outcome="fail" reason="Timecode MUST be a Child Element of Cluster but is not present.">
          <value name="Cluster" offset="1568" context="/Segment[1]/Cluster[1]" formatid="0x1F43B675">[472989 bytes]</value>
        </test>
````
Full report here: https://gist.github.com/kieranjol/cd233abae018ceb05c0cd936a1f3371c
Output file here: https://drive.google.com/file/d/0ByvCvU9YVB09WTNSQ2ZEdXk2YXM/view?usp=sharing

* I see that the colour-matrix option doesn't use the terminology in the Matroska spec, which is `MatrixCoefficients`

That's pretty much it from me, thanks again Moritz.

**Comment by mbunkus** [comment]
Thanks for the feedback.

> According to mediaconch, the file that was generated is not a valid Matroska file. It seems the element in question is:

The file wasn't created by mkvmerge but by `libavf` which is `libavformat` from `ffmpeg` or something using `libavformat` if I'm not mistaken (see mkvinfo's output regarding "multiplexing application" and "writing application". mkvpropedit only modifies the track headers but not the clusters. Therefore this isn't my bug, but ffmpeg's.

> I see that the colour-matrix option doesn't use the terminology in the Matroska spec, which is MatrixCoefficients

That is… huh… that's an accident. I'll change the option's name to `matrix-coefficients`.

Basically my naming scheme is to take the last component from the `element`'s `path` attribute and to convert it to using dashes instead of capitalization. This has obviously gone wrong for `MatrixCoefficients`.

**Comment by mbunkus** [comment]
Hmm now I know why I've used `colour-matrix`. The C++ class name is `KaxColourMatrix`, and I seem to have derived it from the class name. I'll change it to `colour-matrix-coefficients`. I'm using the `colour-` prefix for several properties in order to make them less ambiguous and future-proof (e.g. I cannot simply use `bits-per-channel` as that would clash with audio tracks, obviously).

**Comment by kieranjol** [comment]
Thanks for getting back so quick.

It was indeed made by ffmpeg, however the source file is valid:
https://drive.google.com/open?id=0B80EiSambHhLSnMwQnc1NFZ0Mnc

but the file that was edited by mkvpropedit is not:
https://drive.google.com/file/d/0ByvCvU9YVB09WTNSQ2ZEdXk2YXM/view?usp=sharing

so i was thinking that it must have been something about the `mkvpropedit` process that caused the issue?

Also `colour-matrix-coefficients` sounds perfect, consistent with the other elements.

**Comment by mbunkus** [comment]
Meh, the option added to mkvmerge is called `--colour-matrix`, and that's where I took option names from, too… That code was contributed by someone else. Guess I should change it as well.

About changing that file. Well, I can guarantee you that mkvpropedit is not touching any of the clusters. And if you run `mkvinfo -v -v mkvpropeditoutput.mkv`, you'll see that there is a `cluster timecode` element in the `cluster` at 1568:

```
|+ Cluster at 1568
| + Cluster timecode: 0.000s at 1586
| + SimpleBlock (key, track number 1, 1 frame(s), timecode 0.000s = 00:00:00.000) at 1589
|  + Frame with size 472960
```

One thing that is different is that mkvpropedit had to relocate the track headers (the `Tracks` element) to the end of the file due to there not being enough space at the front. Of course that `Tracks` element is now referenced from the `Meta Seek` element at the start, so any compliant reader should be able to find the `Tracks` element just fine. But maybe MediaConch has some problems with that? See also [this FAQ entry](https://github.com/mbunkus/mkvtoolnix/wiki/All-tracks-vanished-after-mkvpropedit,-the-header-or-chapter-editors) that describes this in more detail.

So I'm really not sure what MediaConch is complaining about, especially why it's emitting that particular error. You should probably check back with their developers.

**Comment by kieranjol** [comment]
Thank you so much. This has gone beyond my reckoning, so I'll link/copy paste your reply as a bug report in Mediaconch. Great work Moritz.

**Comment by JeromeMartinez** [comment]
> So I'm really not sure what MediaConch is complaining about, especially why it's emitting that particular error.

Definitely a misleading error (the `Cluster` parsing is disabled in the library when it is before `Tracks` so the timecode is not parsed then the conformance checker sees that the timecode is not there because it is actually not parsed :(, so wrong error raised).
I fixed that part, no more error now and I am working on supporting file having `Tracks` after `Cluster` (for the moment I am supporting files which have no need of file seek support i.e. `Tracks` before `Cluster`, I'll add a tag saying that the file is not streamable when `Tracks` is after `Cluster` like I do for MOV).

You may consider a feature similar to `ffmpeg -movflags faststart` in order to make the file streamable (i.e. no need of seek) again by having the `Track` element again at the beginning of the file.

**Comment by mbunkus** [comment]
Thank you, Jerome, that sounds absolutely plausible.

There is no such flag in MKVToolNix:

1. mkvmerge always produces files where the tracks are located in front of the clusters; hence it doesn't need such a flag.
2. mkvpropedit cannot guarantee that it's always possible to place updated tracks before the clusters as several muxers (including ffmpeg's `libavformat` and HandBrake if I remember correctly) don't leave any extra space between tracks and clusters (mkvmerge always writes a ~1KB void element there in order to facilitate track header editing without having to relocate the tracks when they grow). The only way to guarantee that would be to rewrite pretty much everything if the tracks grow beyond how much space there is for them at the start of the file — and in that case remuxing with mkvmerge is better anyway.

I may think about a flag to mkvpropedit that aborts file modification if track headers cannot be placed at the start if that's something users might find useful. In such a case the file wouldn't be modified, and it would be left to the user to decide what to do.

**Comment by kieranjol** [comment]
Does `mkclean` do this kind of re-ordering? https://www.matroska.org/downloads/mkclean.html

**Comment by JeromeMartinez** [comment]
> and in that case remuxing with mkvmerge is better anyway.

I didn't think of that, right it may be enough for most people.
I am just aware of some people trying to keep the same structure as the original file when they edit / add metadata i.e. no remuxing (it changes lot of things), but I guess it is a too much specific need for being inside this ticket, and such person would open a dedicated ticket if he needs that.

**Comment by mbunkus** [comment]
> Does mkclean do this kind of re-ordering?

Possibly. Summoning @robUx4 Steve, does mkclean do that?

**Comment by robUx4** [comment]
If you use `--remux` yes.
